🚀 1단계 - 로그인 구현

junseong94 opened the pull request:
안녕하세요 리뷰어님 이번미션 잘 부탁드립니다!

본 미션 관련하여 변경한 내용은 아래를 참고해주시면 감사하겠습니다!!
- reservation 테이블에 theme_id 컬럼이 추가
  - Reservation 엔티티의 ReservationTheme 엔티티 필드 추가
  - ReservationDto 수정 및 ReservationRepository 연관 쿼리 수정

- 이용가능 테마, 시간 조회 API 구현
  - ReservationController 에 "/times/available" GetMapping 구현 
  - ReservationRepository - getAvailableReservationTimes 쿼리 추가
  - 변경 내용에 맞게 테스트 코드 수정
 
- 로그인 기능 구현
  - member 디렉토리 내 구현
  - JWT 의존성 추가
  - 권한 관련 Exception 추가
  - 회원 저장소는 Map 을 이용해 인메모리 형식으로 구현
  - JWT 토큰 관련 코드는 Spring-learning-test 를 참고함.
  - 회원 id 는 atomicLong을 사용
  - 기능 구현 후 테스트 코드 작성

- 미션 진행 중 발생한 이슈
  - reservation 테이블에 themeId 칼럼이 추가되면서 해당 JdbcTemplate 쿼리도 수정하였는데, Reservation 엔티티에는 ReservationTheme 를 필드로 추가하지 않아 해결하는데 오랜 시간이 걸렸었습니다.
    - 이와 관련하여, Reservation과 같은 중요 엔티티에 DB 스키마가 변경되는 것마다 엔티티를 수정하는게 맞나 고민이 되었습니다.
    - DB 스키마가 변경되었다면, 우선 Map과 같은 자료구조로 데이터를 받아와 처리해야할지 그에 맞는 신규 엔티티를 생성할지, 기존 엔티티를 수정해야하는지에 대해 조언 부탁드립니다.


이외에 구조적, 문법적, 기능 사용 등 모든 부분에서 피드백 주시면 감사하겠습니다!

junseong94 committed:
project init

junseong94 committed:
validation 의존성 추가

junseong94 committed:
사용자가 직접 하는 예약 기능 구현
1. 기존 Reservation 엔티티에 themeId 필드 추가로 인한 연관 코드 수정
2. themeId 추가에 따른 테스트 코드 수정

junseong94 committed:
1. 사용자가 직접 예약하는 페이지 연동 - 완료
2. 예약 가능 시간대 응답 API 구현 - 완료
3. 로그인 기능 구현 - 완료
4. 쿠키로 사용자 정보조회 API rngus - 완료

pkeugine commented on src/test/java/roomescape/reservationTime/ReservationTimeRepositoryTest.java:
```diff
@@ -0,0 +1,107 @@
+        assertThat(reservationTimeRepository.existsById(id)).isFalse();
+    }
+
+    @DisplayName("예약 시간을 조회할 수 있습니다")
```
`예약 시간을 조회할 수 있습니다` 라는 이름의 테스트가 중복되는데,
조금 더 구체적으로 이 두 테스트가 어떤 차이가 있는지 나타내보면 좋을 것 같아요.
그리고 뭔가 사소해보이지만 테스트 이름 형태를 비슷하게 맞춰보는건 어떨까요?
존댓말, `@DisplayName` 이 없는 것 등 이런걸 통일하면 관리하는 프로젝트 완성도가 더 높아 보이더라구요.

pkeugine commented on src/main/java/roomescape/member/dto/LoginMemberRequestDto.java:
```diff
@@ -0,0 +1,23 @@
+        return password;
+    }
+
+    public LoginMemberRequestDto(String email, String password) {
+        this.email = email;
+        this.password = password;
+    }
```
보통 생성자를 가장 아래에 두는 편이신가요?
service 와 같은 다른 클래스를 보면 그렇지 않아서요!

junseong94 replied:
저는 주로 생성자를 아래에 두는 편입니다. 
해당 부분에 대해 크게 생각해보지 않았으나, 곰곰이 생각해보니 생성자가 위쪽에 위치하는 게 좋을 것 같아요.
dto나 entity는 생성자를 신규로 추가하는 경우가 많아 위쪽에 위치하는게 가독성면에서 좋을 것 같단 생각이 듭니다!

pkeugine commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Member {
+
+    private final static AtomicLong atomicLong = new AtomicLong(1);
```
[많은 개발자들이 참고하는 코딩 컨벤션](https://google.github.io/styleguide/javaguide.html#s4.8.7-modifiers)을 한 번 확인해보고 프로젝트에 적용해봐도 좋을 것 같아요.
modifier (접근제한자, final, static 등...) 는 보통 특정 순서에 맞게 적어주거든요.
```suggestion
    private static final AtomicLong atomicLong = new AtomicLong(1);
```

그리고 atomicLong 의 경우, 이름을 전체 대문자로 할지, lowerCamelCase 로 할지도 컨벤션을 확인해보고 적용해보면 좋을 것 같아요!

pkeugine commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -0,0 +1,58 @@
+        Cookie[] cookies = request.getCookies();
+        String token = extractTokenFromCookie(cookies);
+        MemberResponseDto memberResponseDto = authService.findMemberName(token);
+        log.info("로그인 계정의 Name : " + memberResponseDto.getName());
```
```suggestion
        log.info("로그인 계정의 Name : {}", memberResponseDto.getName());
```

로그는 이렇게도 작성해볼 수 있어요.
넣어줘야할 정보가 많을 때 전체 메시지를 파악하기 좋은 방식이죠!

pkeugine commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -0,0 +1,58 @@
+        log.info("loginMemberID :" + loginMemberRequestDto.getEmail());
+        TokenResponseDto tokenResponseDto = authService.createToken(loginMemberRequestDto);
+
+        Cookie cookie = new Cookie("token", tokenResponseDto.getAccessToken());
+        cookie.setHttpOnly(true);
+        cookie.setPath("/");
+        response.addCookie(cookie);
```
cookie 의 유효기간을 설정해보는건 어떨까요?
일정 시간이 지나면 지워질 수 있도록 하면 보안상 더 안전할 것 같아요.

pkeugine commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -0,0 +1,58 @@
+
+    @PostMapping("/login")
+    public ResponseEntity<Void> login(@Valid @RequestBody LoginMemberRequestDto loginMemberRequestDto, HttpServletResponse response) {
+        log.info("loginMemberID :" + loginMemberRequestDto.getEmail());
```
로깅을 신경써서 적용해주셨네요! 👍 
해당 내용과 `/login/check` 에서 로그인 계정의 Name 을 남겨주는 이유가 궁금합니다 :)

junseong94 replied:
테스트 코드 작성 과정에서 클라이언트로부터 값이 제대로 들어오는지 확인하기위해 적용했습니다. :)
추후 운영단계 과정에서도 회원정보 관련 요청을 받은 경우 로그로 확인하도록 하는게 유지보수 측면에서도 좋을 거라 생각했습니다.

pkeugine commented on src/main/java/roomescape/reservationTheme/ReservationThemeRequestDto.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.util.Objects;
+
+public class ReservationThemeRequestDto {
```
member 패키지는 dto 를 별도의 디렉토리를 만들어서 관리해주셨는데,
그렇지 않은 도메인도 있어요.
패키지 구조를 비슷하게 유지해보는건 어떨까요?
아니면 해당 방식으로 만들어주신 이유를 말해주셔도 괜찮아요 :)

junseong94 replied:
다른 도메인의 경우 구현을 먼저 신경쓰다보니 도메인 기준으로만 분리했습니다.
Member 도메인 개발시 dto가 명칭들이 다양하다보니 클래스파일 많아져 우선적으로 구분했었는데,
코멘트 주신 이후 전체 도메인 패키지 구조를 수정 완료했습니다!!

pkeugine commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,123 @@
+    public List<ReservationResponseDto> findAll() {
+        List<Reservation> reservations = reservationRepository.findAll();
+        return reservations.stream()
+                .map(
+                        reservation -> new ReservationResponseDto.Builder()
+                                .id(reservation.getId())
+                                .name(reservation.getName())
+                                .date(reservation.getDate())
+                                .reservationTimeResponseDto(new ReservationTimeResponseDto(
+                                        reservation.getReservationTime().getId(),
+                                        reservation.getReservationTime().getStartAt()))
+                                .reservationThemeResponseDto(new ReservationThemeResponseDto(
+                                        reservation.getReservationTheme().getId(),
+                                        reservation.getReservationTheme().getName(),
+                                        reservation.getReservationTheme().getDescription(),
+                                        reservation.getReservationTheme().getThumbnail()
+                                ))
+                                .build()
+                ).toList();
```
```suggestion
        return reservations.stream()
            .map(reservation -> new ReservationResponseDto.Builder()
                .id(reservation.getId())
                .name(reservation.getName())
                .date(reservation.getDate())
                .reservationTimeResponseDto(new ReservationTimeResponseDto(
                    reservation.getReservationTime().getId(),
                    reservation.getReservationTime().getStartAt()))
                .reservationThemeResponseDto(new ReservationThemeResponseDto(
                    reservation.getReservationTheme().getId(),
                    reservation.getReservationTheme().getName(),
                    reservation.getReservationTheme().getDescription(),
                    reservation.getReservationTheme().getThumbnail()
                )).build())
            .toList();
```
엄청 기술적인 내용은 아니지만, 의외로 공백을 관리하는게 가독성에 영향을 많이 준다고 생각해요.
지금 그렇지는 않지만 나중에 로직이 더 복잡하거나 변수 이름들이 길어졌을 때도 영향을 주기도하고요.

(+) 서비스 레이어는 변환 로직을 포함하지 않아도 시간이 갈수록 굉장히 비대해지는 공간이에요.
그러니 이 변환 작업을 다른 곳으로 옮겨보는건 어떨까요?
예를 들어 DTO 에 정적 팩토리 메서드를 만들어볼 수 있겠네요 :)

pkeugine commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,126 @@
+
+    public Reservation findById(Long id) {
+        final String sql = """
+                SELECT r.id as reservation_id, 
+                        r.name as reservation_name, 
+                        r.date as reservation_date, 
+                        t.id as time_id, 
+                        t.start_at as start_at,
+                        th.id  as theme_id,
+                        th.name as theme_name,
+                        th.description as theme_description,
+                        th.thumbnail as theme_thumbnail
+                FROM reservation as r 
+                inner join reservation_time as t on r.time_id = t.id
+                inner join theme as th on r.theme_id = th.id
+                where r.id = ?
```
쿼리의 키워드를 전체 대문자로 사용하기로 했다면 모든걸 다 전체 대문자로 해봐도 괜찮을 것 같아요.

pkeugine commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -0,0 +1,46 @@
+        this.reservationService = reservationService;
+    }
+
+    @GetMapping("/reservations")
+    public ResponseEntity<List<ReservationResponseDto>> getReservations() {
+        final List<ReservationResponseDto> reservationResponseDtos = reservationService.findAll();
+        return ResponseEntity.ok().body(reservationResponseDtos);
+
+    }
+
+    @PostMapping("/reservations")
+    public ResponseEntity<ReservationResponseDto> createReservation(@Valid @RequestBody ReservationRequestDto reservationRequestDto) {
+        final ReservationResponseDto responseDto = reservationService.save(reservationRequestDto);
+        return ResponseEntity.ok().body(responseDto);
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity<Void> deleteReservation(@PathVariable Long id) {
+        reservationService.delete(id);
+        return ResponseEntity.ok().build();
+    }
```
`@RequestMapping` 을 사용하여 api url 의 공통 부분을 추출해보면 어떨까요?

pkeugine commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,8 @@
+# h2 db ??
```
사용하지 않는 주석은 제거해주세요!

(+) application.properties 는 application.yml 형태로도 관리할 수 있어요.
개발자의 취향 차이로 어떤 것을 사용할지 정해지긴하는데,
한 번 둘 중 어떤 것이 준성님께 좋은지 확인해보고 적용해보는것도 재밌을 것 같아요!

pkeugine commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -0,0 +1,42 @@
+
+    public List<ReservationTimeResponseDto> getTimes() {
+
+        final List<ReservationTime> reservationTimes = reservationTimeRepository.findAll();
+        return reservationTimes.stream()
+                .map(
+                        time -> new ReservationTimeResponseDto(time.getId(), time.getStartAt())
+                ).toList();
```
DB 로부터 데이터를 모두 불러온 뒤, application 에서 필터링을 하는 작업을 해주시네요!
해당 로직의 장단점은 뭘까요?

junseong94 replied:
**dto와 entity를 나눠 적용할 경우 아래와 같은 장점이 있다고 생각합니다.**

1. 관심사의 분리
  - 계층별로 요구되는 역할을 분리하기에, 코드가 더 명확해지고, 관리에 용이합니다.
2. 유연한 변경에 따른 변경 용이
  - 데이터베이스와 ui 구조 변경 시 각 역할에 맞게 변경하면 되기에 유연하게 대응이 가능합니다.

pkeugine commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -0,0 +1,42 @@
+    public ReservationTimeResponseDto addTime(ReservationTimeRequestDto reservationTimeRequestDto) {
+        final ReservationTime reservationTime = new ReservationTime(reservationTimeRequestDto.getStartAt());
+        final Long savedTimeId = reservationTimeRepository.save(reservationTime);
+        final ReservationTime savedTime = reservationTimeRepository.findById(savedTimeId);
```
이미 application 쪽에 정보가 있는데 한 번 조회를 해주시는 이유가 궁금해요!

junseong94 replied:
JdbcTemplate에서 Spring Data JPA로 기술 변경이 있을 경우를 대비해 
함수명과 반환 타입을 동일하게 구현했습니다.

그래서 데이터를 저장하고 반환된 Id로 다시한번 DB를 조회하여 ui에 반환하도록 로직을 구성했습니다.

pkeugine replied:
그렇군요!
마침 저희도 나중에 JPA 를 도입할 예정이니 실제로 기술 변경을 해보는 경험을 해볼 수 있겠네요 :)

pkeugine commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Transactional
+    public ReservationTimeResponseDto addTime(ReservationTimeRequestDto reservationTimeRequestDto) {
+        final ReservationTime reservationTime = new ReservationTime(reservationTimeRequestDto.getStartAt());
+        final Long savedTimeId = reservationTimeRepository.save(reservationTime);
+        final ReservationTime savedTime = reservationTimeRepository.findById(savedTimeId);
```
로컬 변수에 final 처리를 해주셨네요.
이러면 어떤 장점이 있을까요?
또한, 준성님의 프로젝트를 보면 final 처리가 되어있지 않은 로컬 변수도 많아요.
어떤 기준을 가지고 final 을 붙여주시는지 궁금해요!

junseong94 replied:
final의 역할은 변수, 객체를 상수 처리하여 혹시 모를 값변경을 예방하기위해 사용한다고 생각합니다.

그러나, final 처리를 어떤 기준으로 해야하는지는 고민해봤는데 잘 모르겠어요. :(
해당 메서드 내에서 로직이 완료가 되면 무조건 final을 붙여서 다른 곳에서는 절대 해당값을 수정하지 못하도록 해야하지 않나 싶어요!

pkeugine replied:
기준은 사람들마다, 그리고 팀마다 다를거라 이번엔 준성님 나름의 기준을 한 번 만들어보면 좋을 것 같아요.
이번에 리팩토링 해주시면서 로컬변수에 전체적으로 모두 final 처리를 해주셨네요 👍 
저도 만약 final 처리를 해준다면 통일성 있게 모든 곳에 해주는게 좋다고 생각해요.

pkeugine commented on .gitignore:
```diff
@@ -35,3 +35,7 @@ out/
 
```
커밋 단위와 커밋 메시지를 한 번 신경써서 관리해보는 것도 괜찮을 것 같아요.
실제 프로젝트를 만들어가는 느낌으로 해보는거죠 :)

pkeugine commented on src/main/java/roomescape/member/AuthService.java:
```diff
@@ -0,0 +1,53 @@
+
+    public void checkMember(Member member) {
+
+        boolean existByEmail = memberRepository.isExistByEmail(member);
+        if (!existByEmail) {
+            throw new AuthorizationException("해당하는 회원 정보가 없습니다.");
+        }
+
+        boolean exsitByEmailAndPassword = memberRepository.isExistMemBerByEmailAndPassword(member.getEmail(), member.getPassword());
+        if (!exsitByEmailAndPassword) {
+            throw new AuthorizationException("해당하는 회원 정보가 없습니다.");
+        }
```
이렇게 두 번 검사해주시는 이유가 궁금해요!
두번째 로직만 있어도 괜찮을 것 같기도 해서요.

junseong94 replied:
그렇네요 
회원정보 제약조건을 생각하다 동일한 이메일있을 수도 있으니, 비밀번호를 추가했었는데 
위에 조건은 생략해도 될 것 같아요.

조언 감사합니다 :)

pkeugine commented on src/main/java/roomescape/member/AuthService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public void checkMember(Member member) {
+
```
어떤 경우엔 메서드 첫 줄이 공백인데,
혹시 이것에 대한 기준도 있으신가요 ㅎㅎ

junseong94 replied:
개인적으로 함수 생성시 첫줄을 항상 띄어놓는게 가독성이 좋다고 생각했었습니다.
개인적인 취향이었는데, 이후에 첫 줄 공백을 삭제해도 코드 가독성은 동일한 것 같아 첫줄 공백없는 convention으로 통일했습니다.

pkeugine commented on src/main/java/roomescape/member/AuthorizationException.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
```
`@ResponseStatus` 와 `@ExceptionHandler`, 이 둘의 역할은 무엇일까요?
둘 다 있어야할까요? 아니면 하나만 있어도 괜찮을까요?
무엇이 더 우선순위로 적용될까요?
모든 질문에 대한 대답을 해주시진 않아도 괜찮지만, 둘이 어떤 역할을 하는지 한 번 알아봐도 괜찮을 것 같아요. :)

junseong94 replied:
아래와 같이 정리해봤습니다. :)

<img width="691" alt="image" src="https://github.com/next-step/spring-roomescape-auth/assets/105576721/5e093105-1fbb-47e3-b229-e8fd2bd985c1">

pkeugine commented on src/main/java/roomescape/member/MemberRepository.java:
```diff
@@ -0,0 +1,42 @@
+
+    public boolean isExistByEmail(Member member){
+        return MemberStorage.values().stream().anyMatch(
+                m -> m.getEmail().equals(member.getEmail())
```
```suggestion
                member -> m.getEmail().equals(member.getEmail())
```

이왕이면 줄임말 대신 명확한 이름을 사용하면 좋을 것 같네요!

pkeugine commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Test
+    void reservation() {
+        Map<String, Object> params = new HashMap<>();
+        params.put("name", "브라운");
+        params.put("date", "2025-08-05");
+
+        Map<String, Object> time = new HashMap<>();
+        time.put("startAt", "15:40");
+        time.put("id", 1);
+
+        Map<String, Object> theme = new HashMap<>();
+        theme.put("id", 1);
+        theme.put("name", "테마1");
+        theme.put("description", "설명1");
+        theme.put("thumbnail", "썸네일1");
+
+        params.put("reservationTimeRequestDto", time);
+        params.put("reservationThemeRequestDto", theme);
```
직접 만들어주신 dto 를 테스트에서 사용해봐도 좋을 것 같아요.

pkeugine commented on src/test/java/roomescape/member/MemberControllerTest.java:
```diff
@@ -0,0 +1,93 @@
+    void basicLogin(){
+
+        final LoginMemberRequestDto loginMemberRequestDto = new LoginMemberRequestDto(EMAIL, PASSOWORD);
+        var response = RestAssured
```
var 과 명시적인 타입을 번갈아가며 사용하기보단 하나로 통일해보죠!

junseong94 replied:
Response로 통일했습니다 :)

pkeugine commented on src/test/java/roomescape/member/MemberControllerTest.java:
```diff
@@ -0,0 +1,93 @@
+    void basicLogin(){
+
+        final LoginMemberRequestDto loginMemberRequestDto = new LoginMemberRequestDto(EMAIL, PASSOWORD);
+        var response = RestAssured
+                .given().log().all()
+                .body(loginMemberRequestDto)
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .accept(MediaType.APPLICATION_JSON_VALUE)
+                .when().post("/login")
+                .then().log().all().extract();
+
+        assertThat(response.cookie("token")).isNotEmpty();
```
```suggestion
        RestAssured
            .given().log().all()
            .body(loginMemberRequestDto)
            .contentType(MediaType.APPLICATION_JSON_VALUE)
            .accept(MediaType.APPLICATION_JSON_VALUE)
            .when().post("/login")
            .then().log().all()
            .assertThat()
            .cookie("token", notNullValue());
```

RestAssured 는 생각보다 훨씬 많은 기능을 제공해요.
공식 문서도 그리 어렵지 않으니 한 번 확인해보고, 이렇게 응답이나 쿠키를 추출하여 검증하는 방식 외에
바로 검증하는 방법도 도전해보면 재밌을 것 같아요!

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,292 @@
+    private String theme2Name;
+    private String theme2Thumbnail;
+    private String theme2description;
+
+
```
불필요한 공백이 있어요.

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,292 @@
+                .collect(Collectors.toList());
+
+        jdbcTemplate.batchUpdate("INSERT INTO reservation(name, date, time_id, theme_id) VALUES (?,?,?,?)", reservations);
+
+
```
불필요한 공백 2!

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,292 @@
+    @Autowired
+    JdbcTemplate jdbcTemplate;
+
+    @BeforeEach
+    void setUp() {
```
before each 로직이 조금 길어서 어떤 행위를 하는지 알기 어려운 것 같아요.
시도해볼 수 있는 방식이 여러개 있는 것 같은데요, 생각나는 것을 한 번 남겨볼게요 :

- 프로덕션과 테스트 환경에 H2 database 적용하기. -> 꽤 쉬운 방법이라 한 번 도전해봐도 괜찮을 것 같아요.
- `@Sql` 어노테이션 알아보고 적용해보기
- 테스트 fixture 관리 방법을 고민해보고 적용해보기
- test 의 given 부분에서 일부 로직을 관리하고, 공통되는 부분을 메서드로 추출해보기

junseong94 replied:
@Sql 및 테스트 fixture 적용해봤습니다 :)

pkeugine commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,120 @@
+    private String name;
+    private String date;
+    private ReservationTime reservationTime;
+    private ReservationTheme reservationTheme;
```
> reservation 테이블에 themeId 칼럼이 추가되면서 해당 JdbcTemplate 쿼리도 수정하였는데, Reservation 엔티티에는 ReservationTheme 를 필드로 추가하지 않아 해결하는데 오랜 시간이 걸렸었습니다. 이와 관련하여, Reservation과 같은 중요 엔티티에 DB 스키마가 변경되는 것마다 엔티티를 수정하는게 맞나 고민이 되었습니다. DB 스키마가 변경되었다면, 우선 Map과 같은 자료구조로 데이터를 받아와 처리해야할지 그에 맞는 신규 엔티티를 생성할지, 기존 엔티티를 수정해야하는지에 대해 조언 부탁드립니다.

보통은 테이블에 컬럼을 추가하면서 entity 에 필드를 함께 추가합니다! (적어도 제가 개발할 때는 그런 것 같아요 🤔)
중요한 entity 이므로, DB 에 변경이 생긴다면 그에 맞게 동일한 변경이 있어야하는 것 같아요.
신규 엔티티를 생성하기엔 변경이 많으면 그만큼 관리해야하는 엔티티라던지, deprecated 된 엔티티가 많아져서
관리하기 오히려 조금 힘들 것 같다는 생각이 드네요.
Map 으로 정보를 가지고 오는 방법은 제가 정확히 이해하지 못했는데, 조금 더 구체적으로 알려주시면 이해하는데 도움이 될 것 같습니다. :)

junseong94 replied:
조언 감사합니다.

Map 적용은 아래 경우와 같이 별도 Entity를 생성하지 않고 Object로 매핑하는 경우입니다!

```java
public Map<String, Object> findUserById(Long id) {
        String sql = "SELECT * FROM user WHERE id = ?";
        return jdbcTemplate.queryForMap(sql, id);
}
```

pkeugine replied:
아 그렇군요!
그렇다면 확실히 앞서 말한 것처럼 entity 를 변경하는 방식이 좋은 것 같아요.
중요한 entity 라 하더라도 잦은 변경이 생기는게 딱히 어색하진 않아요.
오히려 DB 가 아닌, 외부 요소와의 응집성 때문에 잦은 변경이 생긴다면 그건 구조를 다시 설계하거나 리팩토링을 해야한다는 신호가 될 것 같네요 🤔

pkeugine reviewed (changes requested):
안녕하세요 준성님!
이번 미션을 함께할 박유진이라고 합니다 👋 
미션 잘 진행해주셨고, 몇 가지 리뷰 한 번 남겨보았으니 한 번 확인해주세요.
일단은 한 번 request changes 로 가고
원하시는 리뷰를 반영한 뒤 다시 확인해보도록 하죠!
바쁘실텐데 같이 파이팅해보아요! 💪

junseong94 committed:
style : 패키지 구조, yml, log 적용 방식 변경

- application.properties 에서 application.yml 으로 문서 변경
- ui, dto, domain, application, infra 로 패키지 구조 적용
- log.info("Name : {}", 변수) 방식으로 변경

refactor: 하기 항목 리팩토링

- @RequestMapping 적용
  - /times, /login, /reservations, /themes 공통 url을 @RequestMapping으로 추출
  - 전체 Controller에 @RequestMapping 적용
  - AvailableTimesController 클래스 추가
- Reservation entity to dto 변경 함수를 dto class에 추가해 반복되는 코드 생략
  - reservationResponseDtoFromReservation()
- 테스트 코드 반환타입 변경
  - var to final Response
- policy를 bean으로 관리하지 않고, entity에서 static으로 선언
  - ReservationTimeRepository 코드 수정

chore: 동일한 convention 적용을 위한 코드 수정

- 불필요한 공백제거
- dto, entity의 생성자, Getter 순서 변경
- 필드 선언 시 static final로 통일(Member.class)
- 테스트코드 DisplayName 존댓말로 통일
- given, when, then 주석 적용
- 체이닝 함수 convention 수정
- SQL 쿼리 키워드 대문자로 통일
- 테스트 코드 각 변수에 final 적용

junseong94 committed:
refactor: MissionStepTest 테스트코드 리팩토링

- page 테스트에 @DisplayName 추가
- reservation 테스트에 dto 적용

junseong94 committed:
refactor : 불필요햔 조건식 삭제

- 이미 이메일과 패스워드를 검증하고 있는 로직이 있기 대문에 existByEmail 이메일 검증 조건식 삭제

junseong94 committed:
refactor : @ResponseStatus 삭제

- GlobalExceptionHandler에서 @ExceptionHandler로 정의했기에 불필요한 어노테이션으로 삭제

junseong94 committed:
refactor : RestAssured .cookie() 적용

- extract()로 추출하는게 아닌 RestAssured .cookie()를 적용하도록 테스트코드 수정

junseong94 committed:
refactor : (long) 형변환 코드 수정

- (long) 1 이 아닌 1L로 수정

junseong94 committed:
refactor : 테스트 코드 리팩토링

- 엔티티와 dto Fixture 적용
- 초기 데이터 설정을 위해 @Sql과 @TestPropertySource 적용
  - test-application.yml 추가
  - test-schema.sql.sql 및 test-data.sql 추가

junseong94 committed:
refactor : ReservationTheme 테스트 코드 리팩토링

- 엔티티와 dto Fixture 적용
- 초기 데이터 설정을 위해 @Sql과 @TestPropertySource 적용
  - test-application.yml 추가
  - test-schema.sql.sql 및 test-data.sql 추가

junseong94 committed:
refactor : @Autowired jdbcTemplate 삭제

- Fixture 및 @SQL, @TestPropertySource 적용으로 jdbcTemplate 의존성 주입 코드 삭제

junseong94 committed:
refactor : ReservationTime 테스트 코드 리팩토링

- 엔티티와 dto Fixture 적용
- 초기 데이터 설정을 위해 @Sql과 @TestPropertySource 적용
  - test-application.yml 추가
  - test-schema.sql.sql 및 test-data.sql 추가

junseong94 committed:
fix : 전체 통합테스트 시 포트 충돌 수정

- SpringBootTest.WebEnvironment.RANDOM_PORT 적용
- @LocalServerPort 및 RestAssured.port에 해당 포트 할당

junseong94 committed:
refactor : 변수 final 적용

- 수정되지 않는 변수 final 적용

pkeugine commented on src/main/java/roomescape/member/application/AuthService.java:
```diff
@@ -0,0 +1,50 @@
+        this.memberRepository = memberRepository;
+    }
+
+    public TokenResponseDto createToken(final LoginMemberRequestDto loginMemberRequestDto) {
+        final Member loginMember = new Member(loginMemberRequestDto.getEmail(), loginMemberRequestDto.getPassword());
+        checkMember(loginMember);
+        final Long id = memberRepository.findIdByEmail(loginMember.getEmail());
+
+        String accessToken = jwtTokenProvide.createToken(String.valueOf(id));
+        return new TokenResponseDto(accessToken);
+    }
```
현재 준성님께서 유지하시는 컨벤션이
service 의 비즈니스 로직과 controller 안에서는 파라미터와 로컬변수에 모두 final 처리를 하는 것이 맞나요?
도메인이나 dto, 그리고 repository 의 일부분에는 적용되지 않은 것 같아서 궁금했어요.
(바꾸자는 말은 아니고, 정말 궁금해서 남기는 질문입니다!)

eottabom replied:
먼저, 리뷰어가 아닌데 코멘트 남겨서 미리 죄송합니다!!

`final` 도 재밌는 주제인거 같아서 남겨보아요~ ㅎㅎㅎㅎ
(`final` 에 대해서도 생각해보면 좋을 것 같고, 도움이 되시길 바라며.... )

final keyword 도 의견이 많이 분분한거 같아요
어떤 사람은 불변이라는 것을 명확하게 하기 위해서 final keyword 를 사용하는 것이 더 좋고,
성능에도 도움이 된다 라는 이야기를 하고
어떤 사람은 매개변수 등에 과도하게 final keyword 를 사용하는 것은 가독성을 저해한다는 이야기도 있는 것 같아요!

Java 8 부터는 
final keyword 가 선언되지 않은 변수지만, 값이 재할당 되지 않아 final 와 유사하게 동작하는 
`Effectively final` 이라는 개념이 나오긴했어요 (익명 클래스나, 람다식이 사용된 코드에서 확인 가능할 거예요)

제 경험상으로는 매개 변수에 `final` 이 있으면 오히려 가독성이 떨어지더라구요 ㅎㅎ

(관련 있는 articles 를 추가했는데, 너무 많은가요..ㅋㅋ)
https://www.baeldung.com/java-final-performance
https://medium.com/@aarontharris/final-a-java-keyword-1710fb8f1d8f
https://sg.wantedly.com/companies/bebit/post_articles/366756
https://dev.to/moaxcp/marking-parameters-and-local-variables-as-final-2i4e
https://medium.com/@itsandip1/understanding-javas-effectively-final-variables-61e03a0be93a
https://docs.oracle.com/cd/E19159-01/819-3681/6n5srlhjs/index.html
https://www.baeldung.com/java-effectively-final

junseong94 replied:
서비스와 도메인, 레포지토리는 모두 Final을 적용했고 도메인과 dto에는 적용하지 않았습니다.
뭔가 도메인과 dto는 충분히 값이 바뀔 가능성이 있을것 같아 제외 시키는게 맞다고 생각했습니다!

이부분은 @eottabom 님이 남겨주신 내용 토대로 개인적으로 다시 기준을 잡아보겠습니다

pkeugine replied:
너무 좋습니다 이런 의견 공유!!!!
저는 사실 불변 객체를 만들기 위한 필드에 final 추가 외에는 
다른 곳에는 final 을 붙이지 않습니다.
이유는 너무 단순하게도 일관성을 지키기 매우 어렵고 가독성을 해치기 때문인데요,
이는 제 방식이기 때문에 그냥 참고용으로 보시고, 나중에 준성님의 기준도 이후 스텝에서 눈여겨 보도록 하겠습니다 :)

pkeugine commented on src/main/java/roomescape/reservationTheme/ui/ReservationThemeController.java:
```diff
@@ -0,0 +1,46 @@
+        final ReservationThemeResponseDto theme = reservationThemeService.createTheme(requestDto);
+        final HttpHeaders header = new HttpHeaders();
+        header.add("Location", "/theme/" + theme.getId());
+        return new ResponseEntity<>(theme, header, HttpStatus.CREATED);
```
`ResponseEntity.created()` 를 활용해보는건 어떨까요?
다른 메서드와 통일성을 가져갈 수 있을 것 같아요.

junseong94 replied:
오... 되게 유용하군요 감사합니다 :)

pkeugine commented on src/test/java/roomescape/GloblaFixture/dto/ReservationTimeDtoFixture.java:
```diff
@@ -0,0 +1,10 @@
+package roomescape.GloblaFixture.dto;
```
앞서 대화해봤던 코딩 컨벤션 기억나시나요?
그 컨벤션 내용에는 패키지명에 대한 내용도 있는데요,
한 번 읽어보고 그에 맞게 이름을 바꿔보는것도 괜찮을 것 같아요.
만약 바꾼다면 `globalfixture` 가 되겠네요 :)
(다른 패키지도 확인해주세요!)

pkeugine commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -0,0 +1,84 @@
+                .statusCode(200);
+    }
+
+    @DisplayName("시간, 테마 등록 및 예약 등록, 전체조회, 삭제 테스트를 수행합니다.")
+    @Test
+    void reservation() {
```
controller 와 repository 테스트를 잘 만들어주셨고, 관리도 잘 되는 것 같아요 👍 
해당 MissionStepTest 도 한 번 이름을 바꿔서 관리해주고, 적절한 패키지로 옮겨보는건 어떨까요?
만약 적절한 패키지가 없다면 테스트를 분리해봐도 괜찮을 것 같아요.

pkeugine commented on src/test/java/roomescape/member/MemberControllerTest.java:
```diff
@@ -0,0 +1,98 @@
+import static org.assertj.core.api.Assertions.*;
+import static org.hamcrest.CoreMatchers.notNullValue;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+class MemberControllerTest {
+
+    private static final String EMAIL = "json@email.com";
+    private static final String PASSOWORD = "1234";
+
+    @DisplayName("로그인을 하면 토큰을 응답받습니다.")
+    @Test
+    void basicLogin() {
+        //given
+        final LoginMemberRequestDto loginMemberRequestDto = new LoginMemberRequestDto(EMAIL, PASSOWORD);
+
+        // when
+        RestAssured
+                .given().log().all()
+                .body(loginMemberRequestDto)
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .accept(MediaType.APPLICATION_JSON_VALUE)
+                .when().post("/login")
+                .then().log().all()
+                .assertThat()
+                .cookie("token", notNullValue());
+
```
해당 테스트 이름은 `MemberControllerTest` 인데,
테스트 내용을 보면 Controller 뿐만 아니라 모든 객체를 활용하여 통합테스트를 진행하고 있어요.
테스트 이름을 바꿔보거나, 실제로 Controller 만 가지고 테스트하도록 해보는건 어떨까요?

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@TestPropertySource(locations = "classpath:test-application.yml")
+@Sql(scripts = "/test-schema.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = "/test-data.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
👍 👍 리뷰 반영 잘 해주셨네요!

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+        final ReservationResponseDto responseDto = response.as(ReservationResponseDto.class);
+
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(responseDto.getName()).isEqualTo(request.getName());
+        assertThat(responseDto.getDate()).isEqualTo(request.getDate());
+        assertThat(responseDto.getReservationTimeResponseDto().getStartAt())
+                .isEqualTo(request.getReservationTimeRequestDto().getStartAt());
```
assert 하는 것이 여러개라면 `assertAll` 또는 `assertSoftly` 를 한번 알아보고 둘 중 원하시는걸 적용해보면 좋을 것 같아요.

pkeugine commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+        assertThat(response.jsonPath().getString("date")).isEqualTo("예약일자를 입력해주세요");
+    }
+
+    @DisplayName("특정 일자와 테마 선택 시 기 예약된 시간을 제외한 시간을 반환한다.")
+    @Test
+    void getAvaliableTimes() {
+        // given
+        final String date = "2024-12-25";
+        final Long themeId = 1l;
+
+        final String date2 = "2025-12-31";
+        final Long themeId2 = 2l;
+
+        // when
+        final Response response = RestAssured.given().log().all()
+                .queryParam("date", date)
+                .queryParam("themeId", themeId)
+                .when().get("/times/available")
+                .then().log().all().extract().response();
+
+        final Response response2 = RestAssured.given().log().all()
+                .queryParam("date", date2)
+                .queryParam("themeId", themeId2)
+                .when().get("/times/available")
+                .then().log().all().extract().response();
+
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(response.jsonPath().getList(".", ReservationTimeResponseDto.class)).hasSize(2);
+        assertThat(response2.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(response2.jsonPath().getList(".", ReservationTimeResponseDto.class)).hasSize(3);
+    }
```
하나의 테스트에서 두 번 검증을 해요.
하나만 있어도 괜찮지 않을까요?
만약 괜찮지 않다면, 테스트에 파라미터로 여러 조건을 넣어서 테스트해볼 수는 없을까요?
`ParameterizedTest` 를 한 번 알아보는걸 추천드려요!

pkeugine commented on src/test/java/roomescape/reservationTime/ReservationTimeControllerTest.java:
```diff
@@ -0,0 +1,140 @@
+        final ReservationTimeRequestDto request = ReservationTimeDtoFixture.createReservationTimeRequestDto();
+
+        // when
+        final Response response = RestAssured.given().log().all()
+                .contentType(ContentType.JSON)
+                .body(request)
+                .when().post("/times")
+                .then().log().all().extract().response();
+        final ReservationTimeResponseDto responseDto = response.as(ReservationTimeResponseDto.class);
```
시간을 삭제하는 테스트에서 시간을 생성하는 것은 그리 중요한 내용은 아닌 것 같아요.
when 부분에 넣기보다는 메서드를 추출해서 실제 '삭제' 테스트에 집중할 수 있도록 해보는건 어떨까요?
(해당 리뷰는 제안입니다. 필수로 반영하지는 않아도 괜찮습니다.)

junseong94 replied:
아래와 같이 반영해봤습니다 :)

```java
@DisplayName("시간을 삭제할 수 있습니다.")
@Test
void deleteTime() {
        // given
        final Response response = createTime(ReservationTimeDtoFixture.createReservationTimeRequestDto());
        final ReservationTimeResponseDto responseDto = response.as(ReservationTimeResponseDto.class);

        // when
        final Response response2 = RestAssured.given().log().all()
                .when().delete("/times/" + responseDto.getId())
                .then().log().all().extract().response();

        // then
        assertThat(response2.statusCode()).isEqualTo(HttpStatus.OK.value());
}

private Response createTime(ReservationTimeRequestDto request){
      return RestAssured.given().log().all()
              .contentType(ContentType.JSON)
              .body(request)
              .when().post("/times")
              .then().log().all().extract().response();
}
```

pkeugine replied:
엄청 깔끔해졌네요 👍 👍 👍 👍 👍

pkeugine reviewed (approved):
안녕하세요 준성님!
정말 고생하셨습니다. 많은 부분을 개선해주셨네요!
final 처리도 잘 해주신 점 👍 

몇 가지 추가 리뷰 남겨보았는데,
일단 다음 스텝을 진행하면서 적용하셔도 될 것 같아서 merge 하도록 할게요.
새로운 요구사항 적용해주시면서 한 번 확인해주세요!
조금만 더 파이팅해보죠 🔥 🔥 🔥
